### PR TITLE
Contract to runtime call permissioning

### PR DESCRIPTION
**Pull request by roy-powerfinance**
Added delegated runtime dispatch that supports doughnut

### CODE REVIEW

File: frame/contracts/src/wasm/runtime.rs [hunk 1]
```diff
@@ -657,7 +657,14 @@ define_env!(Env, <E: Ext>,
 		};
 		charge_gas(&mut ctx.gas_meter, ctx.schedule, RuntimeToken::ComputedDispatchFee(fee))?;
 
-		ctx.ext.note_dispatch_call(call);
+		match ctx.ext.doughnut() {
```
Comment (jordy25519) [inline frame/contracts/src/wasm/runtime.rs]: There should be both a `dispatch_call` and a `dispatch_delegated_call`.
The first means contract wants to do something for itself, the second means contract wants to do something for user (if doughnut permits).
At the moment user can change contract intended behaviour by passing a doughnut

**Review by jordy25519 — CHANGES_REQUESTED** [review]
👍 Needs refactor so that there are two types of dispatches

File: frame/contracts/src/wasm/runtime.rs (hunk 1 above)
Comment (cowboy-bebug) [reply frame/contracts/src/wasm/runtime.rs]: Thanks for spotting this!

Comment (cowboy-bebug) [reply frame/contracts/src/wasm/runtime.rs]: Are we introducing duplicated calls by doing this?

File: frame/contracts/src/wasm/runtime.rs [hunk 2]
```diff
@@ -662,6 +662,28 @@ define_env!(Env, <E: Ext>,
...
+		// Charge gas for dispatching this call.
+		let fee = {
+			let balance_fee = <<E as Ext>::T as Trait>::ComputeDispatchFee::compute_dispatch_fee(&call);
+			approx_gas_for_balance(ctx.gas_meter.gas_price(), balance_fee)
+		};
+		charge_gas(&mut ctx.gas_meter, ctx.schedule, RuntimeToken::ComputedDispatchFee(fee))?;
+
+		if let Some(doughnut) = ctx.ext.doughnut() {
+			// Use of intermediate variable to avoid `#[warn(mutable_borrow_reservation_conflict)]`
+			let d = (*doughnut).clone();
```
Comment (cowboy-bebug) [inline frame/contracts/src/wasm/runtime.rs:680]: I don't think we need to verify doughnut at this stage (or do we?) because it's verified at
- `ext_call()` (for contract-contract calls) and
- `Contract::call()` (for runtime-contract calls)

Comment (jordy25519) [reply frame/contracts/src/wasm/runtime.rs:680]: Correct, no verification to do here. This is a `contract-runtime` call, it will be seen like any ordinary extrinsic with a doughnut and will be verified in the targeted runtime method

File: frame/contracts/src/wasm/runtime.rs [hunk 3]
```diff
@@ -662,6 +662,28 @@ define_env!(Env, <E: Ext>,
 		Ok(())
 	},
 
+	ext_delegated_dispatch_call(ctx, call_ptr: u32, call_len: u32) => {
```
Comment (jordy25519) [inline frame/contracts/src/wasm/runtime.rs:667]: This is it 👍. Could use a doc comment

**Review by jordy25519 — COMMENTED** [review]
Add an integration test and then we're good 🍩

**Comment by cowboy-bebug** [comment]
~I can't trigger `verify_contract_to_contract_call` to return error anymore because `Contract::call` checks for `verify_runtime_to_contract` first before actual contract call 😞~

^ The below commit should work around this 😄

File: frame/contracts/src/doughnut_integration.rs [hunk 4]
```diff
@@ -85,12 +86,17 @@ pub struct Test;
 
 #[derive(Clone, Eq, PartialEq, Debug, Encode, Decode)]
 pub struct MockDoughnut {
-	verifiable: bool,
+	runtime_verifiable: bool,
```
Comment (jordy25519) [inline frame/contracts/src/doughnut_integration.rs:91]: Can doc comment these for the next guy 😉

File: frame/contracts/src/doughnut_integration.rs [hunk 5]
```diff
@@ -441,23 +454,23 @@ const CODE_CALLER_CONTRACT: &str = r#"
 fn contract_to_contract_call_executes_with_verifiable_doughnut() {
 	let (callee_wasm, callee_code_hash) = compile_module::<Test>(CODE_RETURN_WITH_DATA).unwrap();
 	let (caller_wasm, caller_code_hash) = compile_module::<Test>(CODE_CALLER_CONTRACT).unwrap();
-	let verifiable_doughnut = MockDoughnut::new(true);
+	let verifiable_doughnut = MockDoughnut::new(true, true);
```
Comment (jordy25519) [inline frame/contracts/src/doughnut_integration.rs]: Could builder pattern this but not important

File: frame/contracts/src/doughnut_integration.rs [hunk 6]
```diff
@@ -521,7 +537,163 @@ fn contract_to_contract_call_returns_error_with_unverifiable_doughnut() {
...
+					event: MetaEvent::contract(RawEvent::DelegatedDispatched(DJANGO, verifiable_doughnut, true)),
+					topics: vec![],
+				}
+			]
+		);
+	});
+}
+
+#[test]
+fn delegated_contract_to_runtime_call_returns_error_with_unverifiable_doughnut() {
```
Comment (jordy25519) [inline frame/contracts/src/doughnut_integration.rs]: Name of this test is wrong as this is a test for `verify_runtime_to_contract_call`

File: frame/contracts/src/wasm/mod.rs [hunk 7]
```diff
@@ -1190,6 +1211,44 @@ mod tests {
...
+		)
+	)
+	(func (export "deploy"))
+
+	(data (i32.const 8) "\00\01\2B\00\00\00\00\00\00\00\E5\14\00")
+)
+"#;
+
+	#[test]
+	fn delegated_dispatch_call_works_with_doughnut() {
```
Comment (jordy25519) [inline frame/contracts/src/wasm/mod.rs:1232]: Can add another test for when the doughnut is `None` here, it should error

File: frame/contracts/src/doughnut_integration.rs [hunk 8]
```diff
@@ -521,7 +537,163 @@ fn contract_to_contract_call_returns_error_with_unverifiable_doughnut() {
...
+
+	let (wasm, code_hash) = compile_module::<Test>(CODE_DELEGATED_DISPATCH_CALL).unwrap();
+	let verifiable_doughnut = MockDoughnut::new(true, true);
+	let delegated_origin = RawOrigin::from((Some(DJANGO), Some(verifiable_doughnut.clone())));
+
+	ExtBuilder::default().existential_deposit(50).build().execute_with(|| {
+		Balances::deposit_creating(&ALICE, 1_000_000);
+		Balances::deposit_creating(&DJANGO, 1_000_000);
+
+		assert_ok!(Contract::put_code(Origin::signed(ALICE), 100_000, wasm));
```
Comment (jordy25519) [inline frame/contracts/src/doughnut_integration.rs:587]: I notice some other tests are using `delegated_origin` to `put_code` can you fix them to use Alice like this one (I guess it was left over from when instantiate needed a `Option<Doughnut>`

**Review by jordy25519 — CHANGES_REQUESTED** [review]
Few clean ups, otherwise keen to merge this ASAP

File: frame/contracts/src/doughnut_integration.rs (hunk 5 above)
Comment (cowboy-bebug) [reply frame/contracts/src/doughnut_integration.rs]: Thanks for this comment! It's very clear how it's constructed now! should follow this pattern as much as possible 😄

File: frame/contracts/src/doughnut_integration.rs (hunk 6 above)
Comment (cowboy-bebug) [reply frame/contracts/src/doughnut_integration.rs]: renaming to `delegated_runtime_to_contract_call_returns_error_with_unverifiable_doughnut`

File: frame/contracts/src/wasm/mod.rs (hunk 7 above)
Comment (cowboy-bebug) [reply frame/contracts/src/wasm/mod.rs:1232]: `delegated_dispatch_call_fails_without_doughnut` is added

File: frame/contracts/src/doughnut_integration.rs [hunk 9]
```diff
@@ -441,23 +458,25 @@ const CODE_CALLER_CONTRACT: &str = r#"
 fn contract_to_contract_call_executes_with_verifiable_doughnut() {
 	let (callee_wasm, callee_code_hash) = compile_module::<Test>(CODE_RETURN_WITH_DATA).unwrap();
 	let (caller_wasm, caller_code_hash) = compile_module::<Test>(CODE_CALLER_CONTRACT).unwrap();
-	let verifiable_doughnut = MockDoughnut::new(true);
+	let verifiable_doughnut = MockDoughnut::default()
+		.set_runtime_verifiable(true)
```
Comment (jordy25519) [inline frame/contracts/src/doughnut_integration.rs:462]: Too good 🔥

**Review by jordy25519 — APPROVED** [review]
🚀 Looks good!
